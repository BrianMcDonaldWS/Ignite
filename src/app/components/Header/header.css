.icon {
  padding-left: 15px;
}

.nav {
  z-index: 2;
}

.container {
  @media screen and (min-width: 1024px) {
    display: flex;
    justify-content: space-between;
    width: 100%;
  }

  @media screen and (min-width: 1500px) {
    max-width: 1344px;
    width: 1344px;
    margin: auto;
  }
}

.title {
  font-size: 20px;
  padding: 0;
  padding-right: 50px;
}

.titleText {
  padding: 0;
}

.titleText:only-child {
  padding-left: 1rem;
}

:global .navbar {
  background-color: $APP_COLOR;
  border-radius: 0;

  .navbar-menu {
    border-radius: 0;
  }

  .navbar-menu.is-active .navbar-item {
    text-align: center;
  }
}
