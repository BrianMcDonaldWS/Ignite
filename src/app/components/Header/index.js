import path from 'path';
import React, { Component } from 'react';
import makeClass from 'classnames';
import PropTypes from 'prop-types';
import ReactRouterPropTypes from 'react-router-prop-types';
import { Link } from 'react-router-dom';

import Search from '../Search';
import Icon from '../Icon';
import styles from './header.css';

const index = process.env.index.replace('.md', '.html');

const makeRouterLink = link => {
  if (link === '/') {
    return `/${index}`;
  }

  if (!link.includes(index)) {
    return `/${link}/${index}`;
  }

  return `/${link}`;
};

const NavItem = ({ item: [key, item], ...props }) => {
  const otherPaths = Object.values(props.navItems).filter(val => val !== '/');

  let isActive;

  if (
    (item !== '/' && props.location.pathname.includes(item)) ||
    (props.location.pathname === '/' && props.navItems.root === item) ||
    (item === '/' &&
      !otherPaths.find(path => props.location.pathname.includes(path)) &&
      !props.location.pathname.includes('blog/'))
  ) {
    isActive = true;
  }

  return (
    key !== 'root' && (
      <Link
        key={key}
        className={makeClass('navbar-item', isActive && 'is-active')}
        to={makeRouterLink(item)}
      >
        {key}
      </Link>
    )
  );
};

const GithubLink = ({ githubURL }) =>
  githubURL ? (
    <a
      className="navbar-item"
      href={githubURL}
      rel="noopener noreferrer"
      target="_blank"
    >
      GitHub
      <Icon className={styles.icon} type="fab" icon="github" />
    </a>
  ) : null;

const hasBlogLink = () =>
  Object.values(window.configuration.markdown).find(([page]) =>
    page.includes('blog/')
  );

const BlogLink = ({ className }) =>
  hasBlogLink() ? (
    <Link
      className={makeClass('navbar-item', className)}
      to={path.join(process.env.baseURL, '/blog/index.html')}
    >
      Blog
      <Icon className={styles.icon} type="fas" icon="rss" />
    </Link>
  ) : null;

BlogLink.propTypes = {
  className: PropTypes.string
};

BlogLink.defaultProps = {
  className: null
};

const DocsLink = ({ className }) => (
  <Link
    className={makeClass('navbar-item', className)}
    to={path.join(process.env.baseURL, index)}
  >
    Docs
    <Icon className={styles.icon} type="fas" icon="book" />
  </Link>
);

DocsLink.propTypes = {
  className: PropTypes.string
};

DocsLink.defaultProps = {
  className: null
};

GithubLink.propTypes = {
  githubURL: PropTypes.string
};

GithubLink.defaultProps = {
  githubURL: null
};

class Header extends Component {
  state = {
    menuOpen: false
  };

  closeMenu = () => {
    this.setState({
      menuOpen: false
    });
  };

  onClickHamburger = () => {
    const { menuOpen } = this.state;

    this.setState({
      menuOpen: !menuOpen
    });
  };

  render() {
    return (
      <nav
        className={makeClass('navbar', styles.nav)}
        role="navigation"
        aria-label="main navigation"
      >
        <div className={styles.container}>
          <div className="navbar-brand">
            <Link
              to={path.join(process.env.baseURL, '/home.html')}
              className={makeClass(styles.title, 'navbar-item')}
            >
              {this.props.logo && (
                <img
                  src={`${
                    this.props.logo.includes('http')
                      ? this.props.logo
                      : `./${this.props.logo}`
                  }`}
                  alt="logo"
                  className={makeClass(styles.logo, 'navbar-item')}
                />
              )}
              <span
                className={makeClass(
                  'is-hidden-mobile navbar-item',
                  styles.titleText
                )}
              >
                {this.props.title}
              </span>
            </Link>

            <a
              role="button"
              className={makeClass(
                'navbar-burger',
                this.state.menuOpen && 'is-active'
              )}
              aria-label="menu"
              aria-expanded="false"
              onClick={this.onClickHamburger}
            >
              <span aria-hidden="true" />
              <span aria-hidden="true" />
              <span aria-hidden="true" />
            </a>
          </div>

          <div
            className={makeClass(
              'navbar-menu',
              this.state.menuOpen && 'is-active'
            )}
          >
            <div className="navbar-end" onClick={this.closeMenu}>
              <Search setSearchResults={this.props.setSearchResults} />
              {this.props.navItems ? (
                Object.entries(this.props.navItems).map(item => (
                  <NavItem key={item[0]} item={item} {...this.props} />
                ))
              ) : (
                <DocsLink
                  className={
                    !this.props.location.pathname.includes('blog/') &&
                    !(
                      this.props.location.pathname === '/' &&
                      !window.configuration.markdown['home.md']
                    ) &&
                    'is-active'
                  }
                />
              )}

              <BlogLink
                className={
                  this.props.location.pathname.includes('blog/') && 'is-active'
                }
              />
              <GithubLink githubURL={this.props.githubURL} />
            </div>
          </div>
        </div>
      </nav>
    );
  }
}

Header.propTypes = {
  title: PropTypes.string,
  logo: PropTypes.string,
  githubURL: PropTypes.string,
  navItems: PropTypes.array,
  setSearchResults: PropTypes.func,
  // eslint-disable-next-line react/no-typos
  location: ReactRouterPropTypes.location
};

Header.defaultProps = {
  setSearchResults: () => {},
  title: process.env.title,
  logo: process.env.logo,
  navItems: process.env.navItems,
  githubURL: process.env.githubURL,
  location: {
    pathname: '',
    hash: ''
  }
};

export default Header;
