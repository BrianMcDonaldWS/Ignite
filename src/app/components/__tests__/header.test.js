import React from 'react';
import snapshotDiff from 'snapshot-diff';
import { render, fireEvent, cleanup } from 'react-testing-library';

import Header from '../Header';

afterEach(cleanup);

describe('header', () => {
  beforeAll(() => {
    process.env.baseURL = '/';
    process.env.index = 'index.md';
    window.configuration = { markdown: [] };
  });

  test('renders title', () => {
    const titleText = 'Super Cool Docs';
    const { getByText } = render(<Header title={titleText} />);
    expect(getByText(titleText)).toBeDefined();
  });

  test('renders githubURL', () => {
    const githubURL = 'https://github.com/sindresorhus/cows';
    const { getByText } = render(<Header githubURL={githubURL} />);
    expect(getByText('GitHub')).toBeDefined();
  });

  test('renders logo', () => {
    const logo = 'https://github.com/sindresorhus/cows';
    const { getByAltText } = render(<Header logo={logo} />);
    expect(getByAltText('logo')).toBeDefined();
  });

  test('renders nav items', () => {
    const { getByText } = render(
      <Header
        location={{ hash: '', search: '', pathname: '/' }}
        navItems={{
          root: '/',
          'Doc 1': '/',
          'Doc 2': 'pages'
        }}
      />
    );

    expect(getByText('Doc 1')).toBeDefined();
    expect(getByText('Doc 2')).toBeDefined();
  });

  test('render active for blogs', () => {
    window.configuration = { markdown: [['/blog/page.md']] };
    const { getByText } = render(
      <Header location={{ hash: '', search: '', pathname: '/blog/page.md' }} />
    );
    expect(getByText('Blog')).toBeDefined();
  });

  test('hamburger toggles menu', () => {
    const setSearchResults = jest.fn();
    const { getByLabelText, container: Container1 } = render(
      <Header setSearchResults={setSearchResults} />
    );
    fireEvent.click(getByLabelText('menu'));

    const { getByLabelText: getByLabelText2, container: Container2 } = render(
      <Header setSearchResults={setSearchResults} />
    );

    expect(snapshotDiff(Container1, Container2)).toMatchSnapshot();
    fireEvent.click(getByLabelText2('menu'));
    expect(snapshotDiff(Container1, Container2)).toMatchSnapshot();
  });
});
