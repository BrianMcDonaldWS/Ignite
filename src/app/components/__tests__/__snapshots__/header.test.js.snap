// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clicking hamburger changes state 1`] = `
<nav
  aria-label="main navigation"
  className="navbar"
  role="navigation"
>
  <div
    className={undefined}
  >
    <div
      className="navbar-brand"
    >
      <a
        className="navbar-item"
        href="/"
      >
        <span
          className="is-hidden-mobile"
        />
      </a>
      <a
        aria-expanded="false"
        aria-label="menu"
        className="navbar-burger is-active"
        onClick={[Function]}
        role="button"
      >
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
      </a>
    </div>
    <div
      className="navbar-menu is-active"
    >
      <div
        className="navbar-end"
      >
        <a
          className="navbar-item is-active"
          href="#/"
        >
          Docs
          <i
            className="fas fa-book"
          />
        </a>
      </div>
    </div>
  </div>
</nav>
`;

exports[`renders githubURL 1`] = `
<nav
  aria-label="main navigation"
  className="navbar"
  role="navigation"
>
  <div
    className={undefined}
  >
    <div
      className="navbar-brand"
    >
      <a
        className="navbar-item"
        href="/"
      >
        <span
          className="is-hidden-mobile"
        />
      </a>
      <a
        aria-expanded="false"
        aria-label="menu"
        className="navbar-burger"
        onClick={[Function]}
        role="button"
      >
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
      </a>
    </div>
    <div
      className="navbar-menu"
    >
      <div
        className="navbar-end"
      >
        <DocsLink
          className="is-active"
        />
        <BlogLink
          className={false}
        />
        <GithubLink
          githubURL="https://github.com/sindresorhus/cows"
        />
      </div>
    </div>
  </div>
</nav>
`;

exports[`renders logo 1`] = `
<nav
  aria-label="main navigation"
  className="navbar"
  role="navigation"
>
  <div
    className={undefined}
  >
    <div
      className="navbar-brand"
    >
      <a
        className="navbar-item"
        href="/"
      >
        <img
          alt="logo"
          className="navbar-item"
          src="https://github.com/sindresorhus/cows"
        />
        <span
          className="is-hidden-mobile"
        />
      </a>
      <a
        aria-expanded="false"
        aria-label="menu"
        className="navbar-burger"
        onClick={[Function]}
        role="button"
      >
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
      </a>
    </div>
    <div
      className="navbar-menu"
    >
      <div
        className="navbar-end"
      >
        <DocsLink
          className="is-active"
        />
        <BlogLink
          className={false}
        />
        <GithubLink
          githubURL={null}
        />
      </div>
    </div>
  </div>
</nav>
`;

exports[`renders nav items 1`] = `
<nav
  aria-label="main navigation"
  className="navbar"
  role="navigation"
>
  <div
    className={undefined}
  >
    <div
      className="navbar-brand"
    >
      <a
        className="navbar-item"
        href="/"
      >
        <span
          className="is-hidden-mobile"
        />
      </a>
      <a
        aria-expanded="false"
        aria-label="menu"
        className="navbar-burger"
        onClick={[Function]}
        role="button"
      >
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
      </a>
    </div>
    <div
      className="navbar-menu"
    >
      <div
        className="navbar-end"
      >
        <NavItem
          githubURL={undefined}
          item={
            Array [
              "root",
              "/",
            ]
          }
          location={
            Object {
              "pathname": "/",
            }
          }
          logo={undefined}
          navItems={
            Object {
              "Doc 1": "/",
              "Doc 2": "pages",
              "root": "/",
            }
          }
          title={undefined}
        />
        <NavItem
          githubURL={undefined}
          item={
            Array [
              "Doc 1",
              "/",
            ]
          }
          location={
            Object {
              "pathname": "/",
            }
          }
          logo={undefined}
          navItems={
            Object {
              "Doc 1": "/",
              "Doc 2": "pages",
              "root": "/",
            }
          }
          title={undefined}
        />
        <NavItem
          githubURL={undefined}
          item={
            Array [
              "Doc 2",
              "pages",
            ]
          }
          location={
            Object {
              "pathname": "/",
            }
          }
          logo={undefined}
          navItems={
            Object {
              "Doc 1": "/",
              "Doc 2": "pages",
              "root": "/",
            }
          }
          title={undefined}
        />
        <BlogLink
          className={false}
        />
        <GithubLink
          githubURL={null}
        />
      </div>
    </div>
  </div>
</nav>
`;

exports[`renders nav items with just a index 1`] = `
<nav
  aria-label="main navigation"
  className="navbar"
  role="navigation"
>
  <div
    className={undefined}
  >
    <div
      className="navbar-brand"
    >
      <a
        className="navbar-item"
        href="/"
      >
        <span
          className="is-hidden-mobile"
        />
      </a>
      <a
        aria-expanded="false"
        aria-label="menu"
        className="navbar-burger"
        onClick={[Function]}
        role="button"
      >
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
      </a>
    </div>
    <div
      className="navbar-menu"
    >
      <div
        className="navbar-end"
      >
        <NavItem
          githubURL={undefined}
          item={
            Array [
              "root",
              "/",
            ]
          }
          location={
            Object {
              "pathname": "/foo",
            }
          }
          logo={undefined}
          navItems={
            Object {
              "Doc 1": "other/index.md",
              "Doc 2": "pages",
              "root": "/",
            }
          }
          title={undefined}
        />
        <NavItem
          githubURL={undefined}
          item={
            Array [
              "Doc 1",
              "other/index.md",
            ]
          }
          location={
            Object {
              "pathname": "/foo",
            }
          }
          logo={undefined}
          navItems={
            Object {
              "Doc 1": "other/index.md",
              "Doc 2": "pages",
              "root": "/",
            }
          }
          title={undefined}
        />
        <NavItem
          githubURL={undefined}
          item={
            Array [
              "Doc 2",
              "pages",
            ]
          }
          location={
            Object {
              "pathname": "/foo",
            }
          }
          logo={undefined}
          navItems={
            Object {
              "Doc 1": "other/index.md",
              "Doc 2": "pages",
              "root": "/",
            }
          }
          title={undefined}
        />
        <BlogLink
          className={false}
        />
        <GithubLink
          githubURL={null}
        />
      </div>
    </div>
  </div>
</nav>
`;

exports[`renders title 1`] = `
<nav
  aria-label="main navigation"
  className="navbar"
  role="navigation"
>
  <div
    className={undefined}
  >
    <div
      className="navbar-brand"
    >
      <a
        className="navbar-item"
        href="/"
      >
        <span
          className="is-hidden-mobile"
        >
          Super Cool Docs
        </span>
      </a>
      <a
        aria-expanded="false"
        aria-label="menu"
        className="navbar-burger"
        onClick={[Function]}
        role="button"
      >
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
      </a>
    </div>
    <div
      className="navbar-menu"
    >
      <div
        className="navbar-end"
      >
        <DocsLink
          className="is-active"
        />
        <BlogLink
          className={false}
        />
        <GithubLink
          githubURL={null}
        />
      </div>
    </div>
  </div>
</nav>
`;
